src/read/read_ref.rs

```diff
@@ -0,0 +1,114 @@
+use alloc::vec::Vec;
+use core::convert::TryInto;
+use core::{mem, result};
+
+use crate::pod::{from_bytes, slice_from_bytes, Pod};
+
+type Result<T> = result::Result<T, ()>;
+
+/// A trait for reading references to `Pod` types from a block of data.
+///
+/// This allows parsers to handle both of these cases:
+/// - the block of data exists in memory, and it is desirable
+///   to use references to this block instead of copying it,
+/// - the block of data exists in storage, and it is desirable
+///   to read on demand to minimize I/O and memory usage.
+///
+/// The methods accept `self` by value because `Self` is expected to behave
+/// similar to a reference: it may be a reference with a lifetime of `'a`,
+/// or it may be a wrapper of a reference.
+///
+/// The `Clone` and `Copy` bounds are for convenience, and since `Self` is
+/// expected to be similar to a reference, these are easily satisfied.
+///
+/// Object file parsers typically use offsets to locate the structures
+/// in the block, and will most commonly use the `*_at` methods to
+/// read a structure at a known offset.
+///
+/// Occasionally file parsers will need to treat the block as a stream,
+/// and so convenience methods are provided that update an offset with
+/// the size that was read.
+//
+// An alternative would be for methods to accept `&mut self` and use a
+// `seek` method instead of the `offset` parameters, but this is less
+// convenient for implementers.
+pub trait ReadRef<'a>: Clone + Copy {
+    /// The total size of the block of data.
+    fn len(self) -> Result<u64>;
+
+    /// Get a reference to a `u8` slice at the given offset.
+    ///
+    /// Checks that the offset and size can be satisfied.
+    fn read_bytes_at(self, offset: u64, size: u64) -> Result<&'a [u8]>;
+
+    /// Get a reference to a `u8` slice at the given offset, and update the offset.
+    ///
+    /// Errors include if the offset or size are outside the block of data.
+    fn read_bytes(self, offset: &mut u64, size: u64) -> Result<&'a [u8]> {
+        let bytes = self.read_bytes_at(*offset, size)?;
+        *offset = offset.wrapping_add(size);
+        Ok(bytes)
+    }
+
+    /// Get a reference to a `Pod` type at the given offset, and update the offset.
+    ///
+    /// Checks that the offset, size, and alignment can be satisfied.
+    fn read<T: Pod>(self, offset: &mut u64) -> Result<&'a T> {
```

Comment:
You're right that the documentation for `Pod` doesn't mention alignment, but the function `from_bytes` does check alignment. The documentation fix required here is on `from_bytes`, not `Pod`.

This is a bit ugly though, since the `read_bytes` call isn't given anything about the alignment requirement, so in practice it will either have to make sure everything has the maximum possible alignment, or the `unaligned` feature of this crate must be used so that everything has one-byte alignment (which I should document). Maybe the alignment should be passed through to `read_bytes`/`read_bytes_at`, but then `ReadCache` would need to hash on the alignment too. I'd also have to check that nothing in this crate calls `read_bytes_at` directly and then parses aligned types out of that (which normally works for memory mapped files).

Reply:
> Maybe the alignment should be passed through to `read_bytes`/`read_bytes_at`

You can get the alignment via `std::mem::align::<T>()` so it shouldn't require any additional plumbing. Doing the check in `read_at` et al should be possible.